lib/pure/streams.nim

```diff
@@ -403,6 +475,67 @@ when not defined(js):
     result.peekDataImpl = ssPeekData
     result.writeDataImpl = ssWriteData
 
+  type
+    ByteStream* = ref ByteStreamObj ## a stream that encapsulates a seq[byte]
+    ByteStreamObj* = object of StreamObj
+      data*: seq[byte]
+      pos: int
+
+  proc bsAtEnd(s: Stream): bool =
+    var s = ByteStream(s)
+    return s.pos >= s.data.len
+
+  proc bsSetPosition(s: Stream, pos: int) =
+    var s = ByteStream(s)
+    s.pos = clamp(pos, 0, s.data.len)
+
+  proc bsGetPosition(s: Stream): int =
+    var s = ByteStream(s)
+    return s.pos
+
+  proc bsReadData(s: Stream, buffer: pointer, bufLen: int): int =
+    var s = ByteStream(s)
+    result = min(bufLen, s.data.len - s.pos)
+    if result > 0:
+      copyMem(buffer, addr(s.data[s.pos]), result)
+      inc(s.pos, result)
+    else:
+      result = 0
+
+  proc bsPeekData(s: Stream, buffer: pointer, bufLen: int): int =
+    var s = ByteStream(s)
+    result = min(bufLen, s.data.len - s.pos)
+    if result > 0:
+      copyMem(buffer, addr(s.data[s.pos]), result)
+    else:
+      result = 0
+
+  proc bsWriteData(s: Stream, buffer: pointer, bufLen: int) =
+    var s = ByteStream(s)
+    if bufLen <= 0:
+      return
+    if s.pos + bufLen > s.data.len:
+      setLen(s.data, s.pos + bufLen)
+    copyMem(addr(s.data[s.pos]), buffer, bufLen)
+    inc(s.pos, bufLen)
+
+  proc bsClose(s: Stream) =
+    var s = ByteStream(s)
+    s.data = nil
+
+  proc newByteStream*(s: seq[byte] = @[]): ByteStream =
+    ## creates a new stream from the byte sequence `s`.
+    new(result)
+    result.data = s
+    result.pos = 0
+    result.closeImpl = bsClose
+    result.atEndImpl = bsAtEnd
+    result.setPositionImpl = bsSetPosition
+    result.getPositionImpl = bsGetPosition
+    result.readDataImpl = bsReadData
+    result.peekDataImpl = bsPeekData
+    result.writeDataImpl = bsWriteData
+
```

Comment:
Shouldn't we get a ``readBytes`` or something for ``ByteStream`` that returns ``seq[byte]``?